components/datetime/src/pattern/runtime/pattern.rs

```diff
@@ -0,0 +1,45 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use super::super::{reference, PatternItem, TimeGranularity};
+use alloc::vec;
+use zerovec::{ule::AsULE, ZeroVec};
+
+#[derive(Debug, Clone, PartialEq)]
+#[cfg_attr(
+    feature = "provider_serde",
+    derive(serde::Serialize, serde::Deserialize)
+)]
+pub struct Pattern<'data> {
+    #[cfg_attr(feature = "provider_serde", serde(borrow))]
+    pub items: ZeroVec<'data, PatternItem>,
+    pub(crate) time_granularity: TimeGranularity,
+}
+
+impl<'data> From<&'data reference::Pattern> for Pattern<'data> {
+    fn from(input: &'data reference::Pattern) -> Self {
+        Self {
+            items: ZeroVec::from_slice(&input.items),
+            time_granularity: input.time_granularity,
+        }
+    }
+}
+
+impl From<reference::Pattern> for Pattern<'_> {
+    fn from(input: reference::Pattern) -> Self {
+        Self {
+            items: ZeroVec::Owned(input.items.into_iter().map(|i| i.as_unaligned()).collect()),
```

Comment:
Suggestion: you can use `.clone_from_slice()` here as well to reduce some of the boilerplate.  It should do the exact same thing as you've written out.